### ISSUE TITLE
/related Endpoint: Consider changes to permission_callback

### ISSUE THREAD

**Issue by jblz** [Parsely/wp-parsely#667]
Currently, the endpoint is completely unauthenticated.  That's desirable since it's intended to be accessed from the front-end of the site (as the recommended widget currently does). The backing API endpoint is also unauthenticated.

Should we impose a nonce check or any rate limiting or anything else of the sort to attempt to prevent bad actors from levying the proxy endpoint in nefarious ways?

Something else we might consider is including an [`X-Forwarded-For` request header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) that includes the browser's IP address as best as we can discern it (The [`WP_Community_Events::get_unsafe_client_ip()` function](https://github.com/WordPress/wordpress-develop/blob/26bb9fde3db1ace313356a9ede87e8343be15974/src/wp-admin/includes/class-wp-community-events.php#L240) is a good starting place for how to make a best effort to pull this out of the environment in a way that spans host configurations.

For now, this endpoint is opt-in via filter, but this should be considered when and if it's rolled out broadly.

( _Originally brought up here: https://github.com/Parsely/wp-parsely/pull/649#r805227023_ )

**Comment by acicovic** [comment]
As stated above, all our endpoint `permission_callback()` functions are unauthenticated and this is the desired behavior.

1. With the addition of the `/analytics/posts` endpoint, we moved to using abstract classes and methods, which forces every child endpoint to implement its own `permission_callback()` method. This way, we are sure that a proper `permission_callback()` method has been implemented for every endpoint and that we're not firing a common unauthenticated `permission_callback()` by mistake.
2. The Parse.ly API should handle any rate limiting by itself. We're also using cached results which sends a new request only every 5 minutes if I recall correctly from memory, so implementing some kind of rate limiting would not be useful in my opinion.
3. About nonces, if the endpoint could be called by an external bad actor (?), then this could be a good idea. Having said that, in almost every case the attacker would need to know the API secret. As far as I'm aware, the only endpoint that works without it is `/related` which is used for the Recommendations Block and there wouldn't be any sensitive data there. We could consider implementing this though as an extra security feature, just to be even safer.
4. About using headers, I probably wouldn't go that route, with the main reasons being more complexity and potential [privacy/security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For#security_and_privacy_concerns) concerns.
